core/imds/src/main/java/software/amazon/awssdk/imds/internal/AsyncTokenCache.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.imds.internal;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ForkJoinPool;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+
+@SdkInternalApi
+final class AsyncTokenCache implements Supplier<CompletableFuture<Token>> {
+
+    private volatile Token cachedToken;
+    private final Supplier<CompletableFuture<Token>> supplier;
+    private final Object refreshLock = new Object();
+
+    // TODO [DRAFT] add the ability to pass in the execution service
+    private final ExecutorService executorService = ForkJoinPool.commonPool();
+
+    AsyncTokenCache(Supplier<CompletableFuture<Token>> supplier) {
+        this.supplier = supplier;
+    }
+
+    /**
+     * @return an already completed future if the cached value is still fresh. If the cached value was stale, the returned future
+     * that will be completed once the refresh has been completed.
+     */
+    @Override
+    public CompletableFuture<Token> get() {
+        CompletableFuture<Token> returnFuture = new CompletableFuture<>();
+        CompletableFuture.runAsync(() -> refreshIfNeededAndComplete(returnFuture), executorService);
+        return returnFuture;
+    }
+
+    private void refreshIfNeededAndComplete(CompletableFuture<Token> parentFuture) {
+        if (needsRefresh()) {
+            synchronized (refreshLock) {
+                // the value might have been refreshed while waiting on the lock,
+                // if so, refresh it only if it has already expired agan.
+                if (needsRefresh()) {
+                    try {
+                        CompletableFuture<Token> tokenFuture = supplier.get();
+                        CompletableFutureUtils.forwardExceptionTo(parentFuture, tokenFuture);
+                        cachedToken = tokenFuture.get();
```

Comment:
It feels wrong that supplier has its own executor (for returning the future) and we have to have our own executor as well. Could we remove the extra executor? Something like: (disclaimer: I didn't test this)

```java
private List<CompletableFuture<Token>> waitingFutures = new ArrayList<>();
private boolean refreshRunning = false;
private volatile Token cachedToken;

// Guards reads/writes to waitingFutures/refreshRunning and writes to cachedToken
private final Object refreshLock = new Object();

public CompletableFuture<Token> get() {
    // Optimization: read volatile cachedToken once to check needs-refresh and return result instead of
    // reading it twice.
    Token currentValue = cachedToken;
    if (!needsRefresh(currentValue)) {
        return CompletableFuture.completedFuture(currentValue);
    }

    synchronized (refreshLock) {
        // Make sure the value wasn't refreshed while we were waiting for the lock.
        currentValue = cachedToken;
        if (!needsRefresh(currentValue)) {
            return CompletableFuture.completedFuture(currentValue);
        }

        CompletableFuture<Token> result = new CompletableFuture<>();
        waitingFutures.add(result);

        if (!refreshRunning) {
            startRefresh();
        }

        return result;
    }
}

private void startRefresh() {
    supplier.get().whenComplete((token, throwable) -> {
        List<CompletableFuture<Token>> currentWaitingFutures;

        synchronized (refreshLock) {
            // Instead of completing the waiting future while holding the lock, we copy the list reference and
            // release the lock before completing them. This is just in case someone (naughty) is doing something
            // blocking on the complete calls. It's not good that they're doing that, but at least
            // it won't block other threads trying to acquire the lock.
            currentWaitingFutures = waitingFutures;
            waitingFutures = new ArrayList<>();
            refreshRunning = false;
            if (throwable != null) {
                cachedToken = token;
            }
        }

        if (throwable != null) {
            // I assume the supplier has retries built in, so we don't do it here.
            currentWaitingFutures.forEach(f -> f.completeExceptionally(throwable));
        } else {
            currentWaitingFutures.forEach(f -> f.complete(token));
        }
    });

    refreshRunning = true; // After supplier.get(), in case that throws an exception
}
```